tsl/src/remote/dist_txn.c

```diff
@@ -543,6 +545,19 @@ dist_txn_subxact_callback(SubXactEvent event, SubTransactionId mySubid,
 void
 _remote_dist_txn_init()
 {
+	/*
+	 * Disable 2PC if you are running as a standby
+	 *
+	 * It's good for performance since all queries will be read only on the standby
+	 * Additionally we try to assign xids when 2PC is enabled which doesn't work
+	 * on a standby
+	 */
+	if (RecoveryInProgress() && ts_guc_enable_2pc)
+		ereport(ERROR,
+				(errcode(ERRCODE_TS_OPERATION_NOT_SUPPORTED),
+				 errmsg("disable two-phase commit transactions"),
```

Comment:
I am not sure I think it is great to issue this error only when the first distributed transaction is run. Is there a way we can do it much earlier?

Otherwise I think there's a risk that someone will believe they have successfully set up a standby only to find that it later fails.

Another, probably better, option is to just simply not do 2pc when we know it is not needed, even if `ts_guc_enabled_2pc` is TRUE. I guess the GUC doesn't imply that one has to _always_ do 2pc just because it is enabled.

Reply:
@erimatnor I have refactored this now. Please re-check the new code.